libvmi/os/linux/elfparse.c

```diff
@@ -0,0 +1,422 @@
+    }
+
+    return VMI_SUCCESS;
+}
+
+static inline uint32_t elf_hash(const uint8_t* name)
+{
+    uint32_t h = 0, g;
+
+    for (; *name; name++) {
+        h = (h << 4) + *name;
+        if ((g = h & 0xf0000000))
+            h ^= g >> 24;
+        h &= ~g;
+    }
+
+    return h;
+}
+
+static status_t elf_hash_lookup(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    addr_t str_tab,
+    addr_t sym_tab,
+    addr_t hash_tab,
+    const char *name,
+    addr_t *rva)
+{
+    access_context_t _ctx = *ctx;
+    uint32_t hash = elf_hash((uint8_t*) name);
+    uint32_t table[2];
+
+    _ctx.addr = hash_tab;
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(table), table, NULL))
+        return VMI_FAILURE;
+
+    uint32_t current;
+    _ctx.addr += sizeof(table) + sizeof(uint32_t) * (hash % table[0]);
+    if (VMI_FAILURE == vmi_read_32(vmi, &_ctx, &current))
+        return VMI_FAILURE;
+
+    while (current) {
+        struct elf64_sym sym;
+        _ctx.addr = sym_tab + sizeof(struct elf64_sym) * current;
+        if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(struct elf64_sym), &sym, NULL))
+            break;
+
+        _ctx.addr = str_tab + sym.st_name;
+        char *n = vmi_read_str(vmi, &_ctx);
+        if (n && !strcmp(name, n)) {
+            free(n);
+            *rva = sym.st_value;
+            return VMI_SUCCESS;
+        } else if (n)
+            free(n);
+
+        _ctx.addr = hash_tab + sizeof(table) + sizeof(uint32_t) * (table[0] + current);
+        if (VMI_FAILURE == vmi_read_32(vmi, &_ctx, &current))
+            break;
+    }
+
+    return VMI_FAILURE;
+}
+
+static uint32_t elf_hash_max_symbols(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    addr_t hash_tab,
+    uint32_t *max_symbols)
+{
+    access_context_t _ctx = *ctx;
+    uint32_t table[2];
+
+    _ctx.addr = hash_tab;
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(table), table, NULL))
+        return VMI_FAILURE;
+
+    *max_symbols = table[1];
+    return VMI_SUCCESS;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstack-protector"
+
+void
+elfparse_assign_tables(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    const uint8_t * const image,
+    addr_t *hash,
+    addr_t *gnu_hash,
+    addr_t *str,
+    addr_t *sym)
+{
+    struct elf64_ehdr *elf = (struct elf64_ehdr *) image;
+    struct elf64_phdr ph[elf->e_phnum];
+    struct elf64_phdr *h = ph;
+    struct elf64_dyn d = { .d_tag = ~0 };
+    access_context_t _ctx = *ctx;
+
+    _ctx.addr += elf->e_phoff;
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(struct elf64_phdr) * elf->e_phnum, ph, NULL))
+        return;
+
+    /* assumption: there is exactly one PT_DYNAMIC header. */
+    for (; h < ph + elf->e_phnum; h++) {
+        if (h->p_type == PT_DYNAMIC)
+            break;
+    }
+
+    if (h->p_type != PT_DYNAMIC)
+        return;
+
+    for (_ctx.addr = ctx->addr + h->p_vaddr;
+            d.d_tag != DT_NULL;
```

Comment:
I think there should be an upper bound on this loop so that a malformed header can't make this spin excessively.

Reply:
Yeah, I guess checking against malicious input makes sense. We should also consider doing the same for the PE format. I might send a PR for that later.